Security basics

JohnGrey17 opened the pull request.

JohnGrey17 committed:
Added User Entity

JohnGrey17 committed:
added
userDto
userRequestDTo
UserMapper With methods

JohnGrey17 committed:
added changelog configuration
02-create-users-table.yaml

JohnGrey17 committed:
added UserRepository and extends JpaRepository

JohnGrey17 committed:
added AuthenticationController
added UserRegistrationRequestDto
added UserResponseDto

JohnGrey17 committed:
-

JohnGrey17 committed:
implemented UserRegistration Service
Added Registration Exception
added Handler for Registration Exception
Added common path for all EndPoints

andrii-hoienko commented on src/main/java/org/example/bookstore/dto/bookdto/BookRegistrationRequestDto.java:
```diff
@@ -8,7 +8,7 @@
 
 @Data
-public class BookRequestDto {
+public class BookRegistrationRequestDto {
```
I think word `Registration` is not suitable for this case

andrii-hoienko commented on src/main/java/org/example/bookstore/dto/userdto/UserRegistrationRequestDto.java:
```diff
@@ -0,0 +1,29 @@
+    private String lastName;
+    private String shippingAddress;
+
+    @AssertTrue(message = "Password and repeat password must match")
+    public boolean isPasswordMatch() {
+        return password.equals(repeatPassword);
+    }
```
This logic should be implemented using annotation, please read requirement refarully

andrii-hoienko commented on src/main/java/org/example/bookstore/exception/RegistrationException.java:
```diff
@@ -0,0 +1,7 @@
+package org.example.bookstore.exception;
+
+public class RegistrationException extends Exception {
```
Why this exception is checked?

andrii-hoienko commented on src/main/java/org/example/bookstore/model/User.java:
```diff
@@ -0,0 +1,31 @@
+@Setter
+@ToString
+@Table(name = "users")
+public class User {
```
I believe soft delete should be applied to user as well

karma-o replied:
There is no soft delete requirement for the user entity, and in general, we do not usually delete users, and we won't have a delete endpoint for it in this project. But I think you can still leave it as it is if you want

andrii-hoienko commented on src/main/resources/db/changelog/changes/02-create-users-table.yaml:
```diff
@@ -0,0 +1,40 @@
+                  autoIncrement: true
+                  constraints:
+                    primaryKey: true
+                    nullable: false
```
```suggestion

```

andrii-hoienko commented on src/main/resources/db/changelog/db.changelog-master.yaml:
```diff
@@ -1,3 +1,5 @@
   - include:
       file: db/changelog/changes/01-create-books-table.yaml
+  - include:
+      file: db/changelog/changes/02-create-users-table.yaml
```
```suggestion
      file: db/changelog/changes/02-create-users-table.yaml
      
```

karma-o replied:
not resolved, please add an empty new line at the end of file

andrii-hoienko reviewed (changes requested):
Good job, but some improvements required

JohnGrey17 committed:
Added Custom Validator FieldMatch
removed unique constraints for password in liquiBase
made Registration Exception Unchecked
added soft delete to User entity

JohnGrey17 committed:
Added Custom Validator FieldMatch
removed unique constraints for password in liquiBase
made Registration Exception Unchecked
added soft delete to User entity

karma-o commented on src/main/java/org/example/bookstore/exception/RegistrationException.java:
```diff
@@ -0,0 +1,7 @@
+package org.example.bookstore.exception;
+
+public class RegistrationException extends RuntimeException {
```
Actually task requirements imply that it should be a checked exception:
> Add AuthenticationController with the following methods:
public UserResponseDto register(UserRegistrationRequestDto request) throws RegistrationException

karma-o commented on src/main/java/org/example/bookstore/repository/book/BookSpecificationBuilder.java:
```diff
@@ -1,27 +1,30 @@
-        specification = addSpecification(specification, searchParameters.author(), AUTHOR);
-        specification = addSpecification(specification, searchParameters.title(),TITLE);
+        specification = addSpecification(specification, searchParameters.author(),
+                authorSpecificationProvider.getKey());
```
You can also make a constant in `authorSpecificationProvider` public and import it here for direct use like a static method instead of calling a method, but that is up to you

karma-o commented on src/main/java/org/example/bookstore/validator/FieldMatchValidator.java:
```diff
@@ -0,0 +1,37 @@
+
+            return Objects.equals(firstObj, secondObj);
+        } catch (NoSuchFieldException e) {
+            throw new IllegalArgumentException("Field not found: " + e.getMessage());
+        } catch (IllegalAccessException e) {
+            throw new IllegalArgumentException("Illegal access to field: " + e.getMessage());
```
```suggestion
            throw new IllegalArgumentException("Field not found", e);
        } catch (IllegalAccessException e) {
            throw new IllegalArgumentException("Illegal access to field", e);
```

karma-o commented on src/main/java/org/example/bookstore/validator/FieldMatchValidator.java:
```diff
@@ -0,0 +1,37 @@
+    public boolean isValid(Object value, ConstraintValidatorContext context) {
+        Class<?> clazz = value.getClass();
+        try {
+            Field firstFieldValue = clazz.getDeclaredField(firstFieldName);
```
that is not a field value yet, just a `Field` object, same for the other one
```suggestion
            Field firstField = clazz.getDeclaredField(firstFieldName);
```

karma-o commented on src/main/java/org/example/bookstore/validator/FieldMatchValidator.java:
```diff
@@ -0,0 +1,37 @@
+        try {
+            Field firstFieldValue = clazz.getDeclaredField(firstFieldName);
+            firstFieldValue.setAccessible(true);
+            Object firstObj = firstFieldValue.get(value);
```
that would be our value
```suggestion
            Object firstFieldValue = firstField.get(value);
```

karma-o commented on src/main/resources/db/changelog/changes/02-create-users-table.yaml:
```diff
@@ -0,0 +1,46 @@
+                  type: boolean
+                  defaultValueBoolean: false
+                  constraints:
+                    nullable: false
```
```suggestion
                    nullable: false

```

karma-o reviewed (changes requested):
We still need several improvements

JohnGrey17 committed:
removed soft delete variables from User entity
made Registration Exception checked
Fixed BookSpecificationBuilder (call Constants directly from AuthorSpecificationProvider and TitleSpecificationProvider

andrii-hoienko reviewed (approved):
Good job!
